Address parsing gt

Gabe-Torres opened the pull request:
## Overview

This PR adds functionality to Parserator to parse address strings and split them into their parts, like street number, street name, city, and state. It helps the app handle addresses more effectively.

### Demo

<h2>With valid address</h2>

![Screenshot 2024-07-22 at 9 46 10 PM](https://github.com/user-attachments/assets/9d1bb5c0-9782-402a-83e4-4a0fc5304aa6)

![Screenshot 2024-07-22 at 9 46 19 PM](https://github.com/user-attachments/assets/e9332d46-fe93-4ee5-9514-2b559d3ebc0e)

<h2>With invalid address</h2>

![Screenshot 2024-07-22 at 9 45 44 PM](https://github.com/user-attachments/assets/cd24dfda-4b73-425a-a25e-3e55ff9b6ce1)

![Screenshot 2024-07-22 at 9 45 57 PM](https://github.com/user-attachments/assets/b148c2ae-06e3-4911-9a57-3020ef6ccde1)

This curl command sends a request to the /api/parse/ endpoint to parse an address. It shows how to test the API by sending an address like, 123 main st chicago il, and getting back the parsed components.

run: ` docker-compose up`
```python
curl -X GET "http://localhost:8000/api/parse/?address=123%20main%20st%20chicago%20il"
```

### Notes
Quick note: I had to edit the Dockerfile for docker-compose up to work. I was also unable to get the src within parserator_web/templates/parserator_web/index.html <script> tag in the extra_js to read the correct file. The app works as expected when the code is included directly within index.html.

If I had more time, I’d improve error handling for invalid addresses and cover more potential issues. I’d also update the parsing results to make the address parts more readable on the user side, like changing "AddressNumber" to "Address Number" and "StreetName" to "Street Name".

## Testing Instructions

**How to Test This PR:**

1. **Check out the branch:**
    ```bash
    git checkout address-parsing-gt
    ```

2. **Build Docker images:**
    ```bash
    docker-compose build
    ```

3. **Start the app:**
    ```bash
    docker-compose up
    ```

4. **Run the tests:**
    ```bash
    docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app
    ```

5. **Test the API endpoint:**
    - **Send a request to the `/api/parse/` endpoint:**
      ```bash
      curl -X GET "http://localhost:8000/api/parse/?address=123%20main%20st%20chicago%20il"
      ```
        - **Status Code**: `200 OK`
        - **Response Data**:
          - `address_components` should include:
            - `AddressNumber` with value `123`
            - `PlaceName` with value `chicago`
            - `StateName` with value `il`
            - `StreetName` with value `main`

6. **Test error handling:**
    - **Send a request with an invalid address:**
      ```bash
      curl -X GET "http://localhost:8000/api/parse/?address=123%20main%20st%20chicago%20il%20123%20main%20st"
      ```
        - **Status Code**: `400 Bad Request`
        - **Response Data**: Should contain an error message - `Invalid address format`
        
 7. **Happy Path**
 
 ```python
 def test_api_parse_succeeds(client):
    address_string = '123 main st chicago il'
    response = client.get(f'/api/parse/?address={address_string}')
    assert response.status_code == 200
    data = response.json()
    assert 'address_components' in data
    assert 'AddressNumber' in data['address_components']
    assert 'PlaceName' in data['address_components']
    assert 'StateName' in data['address_components']
    assert 'StreetName' in data['address_components']
    assert data['address_components']['AddressNumber'] == '123'
    assert data['address_components']['PlaceName'] == 'chicago'
    assert data['address_components']['StateName'] == 'il'
    assert data['address_components']['StreetName'] == 'main'
```

 8. **Sad Path**
 
 ```python
 def test_api_parse_raises_error(client):
    address_string = '123 main st chicago il 123 main st'
    response = client.get(f'/api/parse/?address={address_string}')
    assert response.status_code == 400
    data = response.json()
    assert 'detail' in data
    assert data['detail'] == 'Invalid address format'
 ```

Gabe-Torres committed:
fix: compose build unable to run

Gabe-Torres committed:
create: views and index view for address parser

Gabe-Torres committed:
chore: clean up

fgregg commented on parserator_web/views.py:
```diff
@@ -14,11 +14,24 @@ class AddressParse(APIView):
+        try:
+            address_components, address_type = usaddress.tag(address)
+        except usaddress.RepeatedLabelError:
+            address_components, address_type = None, "Invalid"
```
we can raise the ParseError here, and it will propagate to to the get method

Gabe-Torres replied:
Good catch! I'll make that change as well. It will make the code cleaner and more uniform.

fgregg commented on parserator_web/templates/parserator_web/index.html:
```diff
@@ -38,5 +38,49 @@ <h4>Parsing results</h4>
+      
+      fetch(`/api/parse/?address=${encodeURIComponent(addressInput)}`)
+        .then(response => {
+            if (!response.ok) {
```
the response will not be okay if there is ParseError right? 

this is preventing your from catching the detailed error message, so you your code on lines 62-64 never get hit.

can you adjust your code so that the alert uses gets the "Invalid address format" from the json if the response is not ok.

Gabe-Torres replied:
I was wondering why I wasn't getting the desired error back. I will make that change 👍🏽

fgregg commented on parserator_web/templates/parserator_web/index.html:
```diff
@@ -38,5 +38,49 @@ <h4>Parsing results</h4>
+            parseType.textContent = data.address_type;
+      
+            for (const [component, tag] of Object.entries(data.address_components)) {
+              const row = addressComponentsTable.insertRow();
```
nice use of insertRow and insertCell

fgregg reviewed (changes requested):
Nice work! 

Really appreciated the screenshots in the PR!

Your error handling isn't quite right, so I have some suggestions for your there!

Gabe-Torres commented:
Thanks for the suggestions @fgregg! I'll have the updates submitted soon

Gabe-Torres committed:
update: parse method

provide consistent error handling
- raise ParseError
- propagate to get method

Gabe-Torres committed:
update: fetch call for error display

In the case of a response that is not ok a user will receive a detailed error message from the API/json

Gabe-Torres commented:
Updated and pushed the changes @fgregg.  
Here is a [short video](https://jam.dev/c/f90af742-a4ee-4308-9879-a2b873182b8e) explaining what was changed and why.

<h3> With invalid address: </h3>

![Screenshot 2024-08-04 at 3 51 31 PM](https://github.com/user-attachments/assets/f4474972-53fe-4d80-b33d-bca436f86b92)

![Screenshot 2024-08-04 at 3 51 40 PM](https://github.com/user-attachments/assets/8aadea8e-87b8-4500-b89d-965820c5a183)

<p> I updated the fetch call so that it now properly returns the response and captures the detailed error if an issue occurs, like an invalid address format.
 
Now, when a user enters an invalid address, they receive a clear and detailed error message from the API in the JSON response, providing helpful feedback that the address format is invalid.

Also, the parse method now raises a ParseError that propagates to the get method—keeping clean and consistent code across the backend.

Thanks for the suggestions! 👍🏽 <p/>

Gabe-Torres committed:
update: error message handling

- Replace browser alert with in-app error message display

Gabe-Torres commented:
I pushed another update and replaced the browser alert with an in-app error message display. The error message is now shown directly on the page with a red background and white text. This improves user experience and makes it easier to read the feedback!
![2024-08-05 00 03 08](https://github.com/user-attachments/assets/edcd1745-a2b3-48be-9234-7cd710a49cad)
